Assertion error in shrinkStoreIfPossible

Hello,

We have a customer of ours that when they run integration tests using h2 database, they get the following assertion error intermittently,

Caused by: java.lang.AssertionError: 95203328 != 65208320
    at org.h2.mvstore.RandomAccessStore.shrinkStoreIfPossible(RandomAccessStore.java:679) ~[h2-2.2.224.jar:2.2.224]

Presumably this should never happen?  Are you able to tell us what could give rise to this assertion error,

Thanks,

Full stack trace....

org.h2.jdbc.JdbcSQLNonTransientConnectionException: The database has been closed [90098-224]
    at org.h2.message.DbException.getJdbcSQLException(DbException.java:690) ~[h2-2.2.224.jar:2.2.224]
    at org.h2.message.DbException.getJdbcSQLException(DbException.java:489) ~[h2-2.2.224.jar:2.2.224]
    at org.h2.message.DbException.get(DbException.java:212) ~[h2-2.2.224.jar:2.2.224]
    at org.h2.engine.SessionLocal.getTransaction(SessionLocal.java:1610) ~[h2-2.2.224.jar:2.2.224]
    at org.h2.engine.SessionLocal.getStatementSavepoint(SessionLocal.java:1620) ~[h2-2.2.224.jar:2.2.224]
    at org.h2.engine.SessionLocal.setSavepoint(SessionLocal.java:858) ~[h2-2.2.224.jar:2.2.224]
    at org.h2.command.Command.executeUpdate(Command.java:248) ~[h2-2.2.224.jar:2.2.224]
    at org.h2.jdbc.JdbcConnection.rollbackInternal(JdbcConnection.java:1441) ~[h2-2.2.224.jar:2.2.224]
    at org.h2.jdbc.JdbcConnection.rollback(JdbcConnection.java:516) ~[h2-2.2.224.jar:2.2.224]
    at com.zaxxer.hikari.pool.ProxyConnection.rollback(ProxyConnection.java:374) ~[HikariCP-3.3.1.jar:?]
    at com.zaxxer.hikari.pool.HikariProxyConnection.rollback(HikariProxyConnection.java) ~[HikariCP-3.3.1.jar:?]
    at net.corda.nodeapi.internal.persistence.DatabaseTransaction.rollback(DatabaseTransaction.kt:105) ~[corda-node-api-4.11.3.jar:?]
    at net.corda.node.services.statemachine.ActionExecutorImpl.executeRollbackTransaction(ActionExecutorImpl.kt:233) ~[corda-node-4.11.3.jar:?]
    at net.corda.node.services.statemachine.ActionExecutorImpl.executeAction(ActionExecutorImpl.kt:67) ~[corda-node-4.11.3.jar:?]
    at net.corda.node.services.statemachine.TransitionExecutorImpl.executeTransition(TransitionExecutorImpl.kt:47) ~[corda-node-4.11.3.jar:?]
    at net.corda.node.services.statemachine.interceptors.HospitalisingInterceptor.executeTransition(HospitalisingInterceptor.kt:39) ~[corda-node-4.11.3.jar:?]
    at net.corda.node.services.statemachine.interceptors.DumpHistoryOnErrorInterceptor.executeTransition(DumpHistoryOnErrorInterceptor.kt:38) ~[corda-node-4.11.3.jar:?]
    at net.corda.node.services.statemachine.FlowStateMachineImpl.processEvent(FlowStateMachineImpl.kt:195) ~[corda-node-4.11.3.jar:?]
    at net.corda.node.services.statemachine.FlowStateMachineImpl.processEventImmediately(FlowStateMachineImpl.kt:301) ~[corda-node-4.11.3.jar:?]
    at net.corda.node.services.statemachine.FlowStateMachineImpl.run(FlowStateMachineImpl.kt:390) ~[corda-node-4.11.3.jar:?]
    at net.corda.node.services.statemachine.FlowStateMachineImpl.run(FlowStateMachineImpl.kt:79) ~[corda-node-4.11.3.jar:?]
    at co.paralleluniverse.fibers.Fiber.run1(Fiber.java:1110) ~[quasar-core-0.7.16_r3.jar:0.7.16_r3]
    at co.paralleluniverse.fibers.Fiber.exec(Fiber.java:806) ~[quasar-core-0.7.16_r3.jar:0.7.16_r3]
    at co.paralleluniverse.fibers.RunnableFiberTask.doExec(RunnableFiberTask.java:103) ~[quasar-core-0.7.16_r3.jar:0.7.16_r3]
    at co.paralleluniverse.fibers.RunnableFiberTask.run(RunnableFiberTask.java:94) ~[quasar-core-0.7.16_r3.jar:0.7.16_r3]
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_382]
    at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_382]
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) ~[?:1.8.0_382]
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) ~[?:1.8.0_382]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_382]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_382]
    at net.corda.node.utilities.AffinityExecutor$ServiceAffinityExecutor$1$thread$1.run(AffinityExecutor.kt:63) ~[corda-node-4.11.3.jar:?]
Caused by: org.h2.mvstore.MVStoreException: java.lang.AssertionError: 95203328 != 65208320 [2.2.224/3]
    at org.h2.mvstore.DataUtils.newMVStoreException(DataUtils.java:996) ~[h2-2.2.224.jar:2.2.224]
    at org.h2.mvstore.FileStore.storeBuffer(FileStore.java:1529) ~[h2-2.2.224.jar:2.2.224]
    at org.h2.mvstore.FileStore.lambda$serializeAndStore$3(FileStore.java:1432) ~[h2-2.2.224.jar:2.2.224]
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_382]
    at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_382]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_382]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_382]
    at java.lang.Thread.run(Thread.java:750) [?:1.8.0_382]
Caused by: java.lang.AssertionError: 95203328 != 65208320
    at org.h2.mvstore.RandomAccessStore.shrinkStoreIfPossible(RandomAccessStore.java:679) ~[h2-2.2.224.jar:2.2.224]
    at org.h2.mvstore.RandomAccessStore.writeChunk(RandomAccessStore.java:377) ~[h2-2.2.224.jar:2.2.224]
    at org.h2.mvstore.RandomAccessStore.writeChunk(RandomAccessStore.java:28) ~[h2-2.2.224.jar:2.2.224]
    at org.h2.mvstore.FileStore.storeBuffer(FileStore.java:1524) ~[h2-2.2.224.jar:2.2.224]
    at org.h2.mvstore.FileStore.lambda$serializeAndStore$3(FileStore.java:1432) ~[h2-2.2.224.jar:2.2.224]
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_382]
    at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_382]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_382]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_382]
    at java.lang.Thread.run(Thread.java:750) ~[?:1.8.0_382]

See the same issue appearing intermittently (but frequently) on an integration-test that runs a series of upgrade SQLs on a database in the integration tests of https://github.com/jeremylong/DependencyCheck during CI builds in Github.

Up to now I've not hit the same case with testruns on my local system.
EDIT: Also happened on my local system, so not isolated to the Github CI environment


```
Caused by: org.h2.mvstore.MVStoreException: java.lang.AssertionError: 189075456 != 165511168 [2.3.232/3]
	at org.h2.mvstore.DataUtils.newMVStoreException(DataUtils.java:996)
	at org.h2.mvstore.FileStore.storeBuffer(FileStore.java:1533)
	at org.h2.mvstore.FileStore.lambda$serializeAndStore$3(FileStore.java:1436)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.AssertionError: 189075456 != 165511168
	at org.h2.mvstore.RandomAccessStore.shrinkStoreIfPossible(RandomAccessStore.java:679)
	at org.h2.mvstore.RandomAccessStore.writeChunk(RandomAccessStore.java:377)
	at org.h2.mvstore.RandomAccessStore.writeChunk(RandomAccessStore.java:28)
	at org.h2.mvstore.FileStore.storeBuffer(FileStore.java:1528)
	... 6 more
```


Failure observed in the tests running under Maven in Github CI action:

```
Apache Maven 3.8.8 (4c87b05d9aedce574290d1acc98575ed5eb6cd39)
Maven home: /usr/share/apache-maven-3.8.8
Java version: 11.0.24, vendor: Azul Systems, Inc., runtime: /opt/hostedtoolcache/Java_Zulu_jdk/11.0.24-8/x64
Default locale: en, platform encoding: UTF-8
OS name: "linux", version: "6.8.0-1014-azure", arch: "amd64", family: "unix"
```

Local system is running
```
Apache Maven 3.9.8 (36645f6c9b5079805ea5009217e36f2cffd34256)
Maven home: /opt/local/share/java/maven3
Java version: 21.0.4, vendor: Eclipse Adoptium, runtime: /Library/Java/JavaVirtualMachines/temurin-21.jdk/Contents/Home
Default locale: en_GB, platform encoding: UTF-8
OS name: "mac os x", version: "14.6.1", arch: "aarch64", family: "mac"
```

I see the following error often on version `2.3.232` of h2, but have no issues with version `2.1.214`:

```
java.lang.AssertionError: 636416000 != 21377024
	at org.h2.mvstore.RandomAccessStore.shrinkStoreIfPossible(RandomAccessStore.java:679)
	at org.h2.mvstore.RandomAccessStore.writeChunk(RandomAccessStore.java:377)
	at org.h2.mvstore.RandomAccessStore.writeChunk(RandomAccessStore.java:28)
	at org.h2.mvstore.FileStore.storeBuffer(FileStore.java:1528)
	at org.h2.mvstore.FileStore.lambda$serializeAndStore$3(FileStore.java:1436)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

jdbc url: `jdbc:h2:~/.h2db1/h2db1;CIPHER=AES;DB_CLOSE_DELAY=-1;CACHE_SIZE=262144`

Apple M1 Max. APFS filesystem.  Sequoia 15.3.2

--- a/h2/src/main/org/h2/mvstore/MVStore.java
+++ b/h2/src/main/org/h2/mvstore/MVStore.java
@@ -910,11 +910,11 @@ public final class MVStore implements AutoCloseable {
         // we need to prevent re-entrance, which may be possible,
         // because meta map is modified within storeNow() and that
         // causes beforeWrite() call with possibility of going back here
-        return !storeLock.isHeldByCurrentThread() || !storeOperationInProgress.get();
+        return !isLockedByCurrentThread() || !storeOperationInProgress.get();
     }
 
     private long store(boolean syncWrite) {
-        assert storeLock.isHeldByCurrentThread();
+        assert isLockedByCurrentThread();
         if (isOpenOrStopping() && hasUnsavedChanges() && storeOperationInProgress.compareAndSet(false, true)) {
             try {
                 @SuppressWarnings({"NonAtomicVolatileUpdate", "NonAtomicOperationOnVolatileField"})
@@ -961,7 +961,7 @@ public final class MVStore implements AutoCloseable {
             int currentUnsavedMemory = unsavedMemory;
             long version = currentVersion;
 
-            assert storeLock.isHeldByCurrentThread();
+            assert isLockedByCurrentThread();
             fileStore.storeIt(collectChangedMapRoots(version), version, syncWrite);
 
             // some pages might have been changed in the meantime (in the newest
@@ -1383,7 +1383,7 @@ public final class MVStore implements AutoCloseable {
                 // condition below is to prevent potential deadlock,
                 // because we should never seek storeLock while holding
                 // map root lock
-                (storeLock.isHeldByCurrentThread() || !map.getRoot().isLockedByCurrentThread()) &&
+                (isLockedByCurrentThread() || !map.getRoot().isLockedByCurrentThread()) &&
                 // to avoid infinite recursion via store() -> dropUnusedChunks() -> layout.remove()
                 fileStore.isRegularMap(map)) {
             saveNeeded = false;
@@ -1818,7 +1818,7 @@ public final class MVStore implements AutoCloseable {
      */
     public void deregisterVersionUsage(TxCounter txCounter) {
         if(decrementVersionUsageCounter(txCounter)) {
-            if (storeLock.isHeldByCurrentThread()) {
+            if (isLockedByCurrentThread()) {
                 dropUnusedVersions();
             } else if (storeLock.tryLock()) {
                 try {
@@ -1839,6 +1839,10 @@ public final class MVStore implements AutoCloseable {
      */
     public boolean decrementVersionUsageCounter(TxCounter txCounter) {
         return txCounter != null &&  txCounter.decrementAndGet() <= 0;
+    }
+
+    boolean isLockedByCurrentThread() {
+        return storeLock.isHeldByCurrentThread();
     }
 
     void onVersionChange(long version) {

--- a/h2/src/main/org/h2/mvstore/RandomAccessStore.java
+++ b/h2/src/main/org/h2/mvstore/RandomAccessStore.java
@@ -157,8 +157,9 @@ public abstract class RandomAccessStore extends FileStore<SFChunk>
     @Override
     protected final boolean validateFileLength(String msg) {
         assert saveChunkLock.isHeldByCurrentThread();
-        assert getFileLengthInUse() == measureFileLengthInUse() :
-                getFileLengthInUse() + " != " + measureFileLengthInUse() + " " + msg;
+        assert !mvStore.isLockedByCurrentThread() || getFileLengthInUse() == measureFileLengthInUse() :
+                        mvStore.isLockedByCurrentThread() + " && " +
+                        getFileLengthInUse() + " != " + measureFileLengthInUse() + " " + msg;
         return true;
     }
 
@@ -676,7 +677,9 @@ public abstract class RandomAccessStore extends FileStore<SFChunk>
     protected void shrinkStoreIfPossible(int minPercent) {
         assert saveChunkLock.isHeldByCurrentThread();
         long result = getFileLengthInUse();
-        assert result == measureFileLengthInUse() : result + " != " + measureFileLengthInUse();
+        assert !mvStore.isLockedByCurrentThread() || result == measureFileLengthInUse() :
+                mvStore.isLockedByCurrentThread() + " && " +
+                result + " != " + measureFileLengthInUse();
         shrinkIfPossible(minPercent);
     }
 

--- a/h2/src/test/org/h2/test/store/SequenceMap.java
+++ b/h2/src/test/org/h2/test/store/SequenceMap.java
@@ -48,7 +48,7 @@ public class SequenceMap extends MVMap<Long, Long> {
 
                     @Override
                     public Long next() {
-                        return Long.valueOf(x++);
+                        return x++;
                     }
 
                 };

--- a/h2/src/test/org/h2/test/synth/TestCrashAPI.java
+++ b/h2/src/test/org/h2/test/synth/TestCrashAPI.java
@@ -236,7 +236,7 @@ public class TestCrashAPI extends TestDb implements Runnable {
             }
             throw e;
         }
-        int len = random.getInt(50);
+        int len = random.getInt(statements.size());
         int first = random.getInt(statements.size() - len);
         int end = first + len;
         Statement stat = conn.createStatement();

